File: NeetCode150/SlidingWindow/LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.java
package NeetCode150.SlidingWindow.LongestSubstringWithoutRepeatingCharacters;

import java.util.HashMap;
import java.util.Map;

public class LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        Map<String, Integer> hashStr = new HashMap<>();
        Integer maxLength = 0;
        Integer leftPointer = 0;
        String[] letters = s.split("");

        if (s.length() == 0) {
            return 0;
        }

        for (int i = 0; i < letters.length; i++) {
            if (hashStr.containsKey(letters[i])) {
                leftPointer = Math.max(leftPointer, hashStr.get(letters[i]) + 1);
            }

            hashStr.put(letters[i], i);
            maxLength = Math.max(maxLength, i - leftPointer + 1);
        }

        return maxLength;
    }
}

// This problem can be solved by using a sliding window.
// We use a hashmap to store the characters and their last index.
// We use a left pointer for the start and right pointer for the end of the window.
// We use a maxLength to keep track of the longest substring without repeating characters.
// Whenever we encounter a character that is already in the hashmap, we update the left pointer to the next character after the last occurrence of the character.
// This solution has O(n) time complexity and O(n) space complexity.

File: NeetCode150/SlidingWindow/LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.py
"""
Given a string s, find the length of the longest substring without repeating characters.

Example 1:
Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.

Example 2:
Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.

Example 3:
Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.
"""

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        strMap = {}
        left = 0
        longest = 0
        
        for right in range(len(s)):
            while s[right] in strMap:
                del strMap[s[left]]
                left += 1
            strMap[s[right]] = True
            longest = max(longest, right - left + 1)
        return longest
      
"""
For this problem, we can use sliding window & hash map for optimal solution. If we meet a duplicate character, we can move the left pointer
to the right by + 1 (slide the substring window to the right) until the duplicate characters is gone. Then, we count the length of window
and store the maximum everytime we add a new character.
"""
